ttax/decompositions.py

```diff
@@ -0,0 +1,251 @@
+import numpy as np
+import jax
+import jax.numpy as jnp
+
+from ttax.base_class import TT
+from ttax.base_class import TTMatrix
+
+
+def round(tt, max_tt_rank=None, epsilon=None):
+  """TT-rounding procedure, returns a TT object with smaller TT-ranks.
+  Args:
+    tt: TT-tensor or TT-matrix.
+    max_tt_rank: a number or a list of numbers.
+      If a number, than defines the maximal TT-rank of the result.
+      If a list of numbers, than `max_tt_rank` length should be d+1
+      (where d is the number of dimensions) and `max_tt_rank[i]` defines
+      the maximal (i+1)-th TT-rank of the result.
+      The following two versions are equivalent
+        `max_tt_rank = r`
+      and
+        `max_tt_rank = [1] + [r] * (d-1) + [1]`
+    epsilon: a floating point number or None
+      If the TT-ranks are not restricted (`max_tt_rank=None`), then
+      the result would be guarantied to be `epsilon` close to `tt`
+      in terms of relative Frobenius error:
+        ||res - tt||_F / ||tt||_F <= epsilon
+      If the TT-ranks are restricted, providing a loose `epsilon` may
+      reduce the TT-ranks of the result.
+      E.g.
+        round(tt, max_tt_rank=100, epsilon=0.9)
+      will probably return you a TT-tensor with TT-ranks close to 1, not 100.
+      Note that providing a nontrivial (= not equal to None) epsilon will make
+      the TT-ranks of the result change depending on the data, 
+      which will prevent you from using jax.jit 
+      for speeding up the computations.
+  Returns:
+    TT-tensor or TT-matrix.
+  Raises:
+    ValueError if max_tt_rank is less than 0, if max_tt_rank is not a number and
+    not a vector of length d + 1 where d is the number of dimensions of
+    the input tensor, if epsilon is less than 0.
+  """
+
+  if max_tt_rank is None:
+    max_tt_rank = np.iinfo(np.int32).max
+  num_dims = tt.ndims
+  max_tt_rank = np.array(max_tt_rank).astype(np.int32)
+  if np.any(max_tt_rank < 1):
+    raise ValueError('Maximum TT-rank should be greater or equal to 1.')
+  if epsilon is not None and epsilon < 0:
+    raise ValueError('Epsilon should be non-negative.')
+  if max_tt_rank.size == 1:
+    max_tt_rank = (max_tt_rank * np.ones(num_dims + 1)).astype(jnp.int32)
+  elif max_tt_rank.size != num_dims + 1:
+    raise ValueError('max_tt_rank should be a number or a vector of size (d+1) '
+                     'where d is the number of dimensions (rank) of the tensor.')
+  if tt.is_tt_matrix:
+    raw_shape = tt.raw_tensor_shape
+  else:
+    raw_shape = tt.shape
+
+  tt_cores = orthogonalize_tt_cores(tt, True).tt_cores
```

Comment:
Either remove True since it's default anyway, or give it a name so it's more readable "left_to_right=True"